cpp/include/cudf/rolling.hpp

```diff
@@ -379,6 +381,119 @@ std::unique_ptr<column> grouped_time_range_rolling_window(
   std::unique_ptr<aggregation> const& aggr,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @brief  Applies a grouping-aware, value range-based rolling window function to the values in a
+ *         column.
+ *
+ * This is a generalization of `grouped_time_range_rolling_window()` to work with non-timestamp
+ * columns. Like `grouped_time_range_rolling_window()`, this function aggregates values in a window
+ * around each element of a specified `input` column. It differs from
+ * `grouped_time_range_rolling_window()` in that instead of a timestamp column, the ordered column
+ * can be of integral types as well as timestamps.
+ *
+ *   1. The elements of the `input` column are grouped into distinct groups (e.g. the result of a
+ *      groupby), determined by the corresponding values of the columns under `group_keys`. The
+ *      window-aggregation cannot cross the group boundaries.
+ *   2. Within a group, the aggregation window is calculated based on an interval (e.g. number
+ *      of days preceding/following the current row). The value intervals are applied on the
+ *      `orderby_column` argument.
+ *
+ * Note: This method requires that the rows are presorted by the group keys and orderby column
+ * values.
+ *
+ * The window intervals are specified as scalar values appropriate for the orderby column:
+ *   1. If `orderby` column is a timestamp, the `preceding`/`following` windows are specified
+ *      in terms of lower resolution `duration` scalars.
+ *      E.g. For `orderby` column of type `TIMESTAMP_SECONDS`, the intervals may be
+ * `DURATION_SECONDS` or `DURATION_DAYS`. Higher resolution durations (e.g. `DURATION_NANOSECONDS`)
+ * cannot be used with lower resolution timestamps.
+ *   2. If the `orderby` column is an integral type (e.g. `INT32`), the `preceding`/`following`
+ *      should be the exact same type (`INT32`).
```

Comment:
This seems inconsistent. Seems like the timestamp interval should also be the same type. There's no loss of information in converting a lower resolution timestamp value to a higher resolution, yes? E.g., if the orderby column is seconds, and the internal is `[-1, +1]` days, then that's no different than `[-86400, +86400]` seconds.

Reply:
I haven't grasped your point, I think. 

The order-by column is the timestamp at a certain resolution. The window bounds are durations (with matching or lower resolution). The bounds can't be timestamps because they are relative intervals. (E.g. 7 days preceding, 86400 seconds following.)

The old API took `size_type` and interpreted them in days (I.e. lowest resolution).

Reply:
> Otherwise you need a double dispatch.

You're right; I'm doing the double-dispatch in this version of things. (By the way, is host-side dispatch expensive? My understanding is that the dispatch itself isn't more expensive than an `if` check. The real cost is in the casting.)

The reason this is permitted is to support how Spark/SQL specifies intervals at different resolutions. Would you recommend having the caller cast appropriately instead?

(The same will likely apply for decimals.)

Reply:
> (By the way, is host-side dispatch expensive? My understanding is that the dispatch itself isn't more expensive than an if check. The real cost is in the casting.)

It's not really expensive in runtime, but can be quite expensive in compile time/binary size depending on the amount of code in the leaves of the double dispatched code paths. 


> The reason this is permitted is to support how Spark/SQL specifies intervals at different resolutions. Would you recommend having the caller cast appropriately instead?

Yes, the way I see it, it doesn't seem any different from how Pandas/Spark allow joining on columns of different types but libcudf doesn't support this. The caller is expected to perform any casting beforehand.

Reply:
Hmm... That'd certainly simplify certain other aspects of this feature.

Let me try pulling out the scaling part.

Reply:
Hey, Jake. If the scaling feature is removed, I'd still need to check that the window-duration and the orderby-timstamps are of the same resolution.
I don't know how I'd do that without doing a double dispatch, short of long if-else.

Would you be averse to the double dispatch in this case?

Reply:
Ok, it took me a while to realize what should have been obvious: The scaling functionality is pre-existing. :/

The `grouped_time_range...()` function already permitted timestamp columns in any resolution, accepted the window widths in `days`, and scaled it to the timetamp's resolution. All that `scale_to()` does is to move the scaling logic into `range_window_bounds`.

Removing `range_window_bounds::scale_to()` would break existing functionality in `grouped_time_range...()`, and callers thereof (including Spark) who depend on the window widths being in `DAYS`.

For what it's worth, there isn't a way to avoid the double-dispatch anyway, even without `scale_to()`, because we'd still have to validate that the window range type (`numeric` or `duration`) are compatible with the order-by column (`numeric` or `timestamp`). This is currently checked as part of the `scale_to` call stack.